mlflow/torchscript.py

```diff
@@ -0,0 +1,108 @@
+from pathlib import Path
+import logging
+import yaml
+
+import torch
+import pandas as pd
+import numpy as np
+import mlflow
+from mlflow import pyfunc
+from mlflow.models import Model
+from mlflow.exceptions import MlflowException
+from mlflow.utils.environment import _mlflow_conda_env
+from mlflow.tracking.artifact_utils import _download_artifact_from_uri
+from mlflow.utils.model_utils import _get_flavor_configuration
+
+FLAVOR_NAME = "torchscript"
+_SERIALIZED_TORCH_MODEL_FILE_NAME = "model.pt"
+_logger = logging.getLogger(__name__)
+
+
+def get_default_conda_env():
+    """
+    :return: The default Conda environment for MLflow Models produced by calls to
+             :func:`save_model()` and :func:`log_model()`.
+    """
+    import torch
+
+    return _mlflow_conda_env(
+        additional_conda_deps=[
+            "pytorch={}".format(torch.__version__),
+        ],
+        additional_conda_channels=[
+            "pytorch",
+        ])
+
+
+def log_model(model, artifact_path, conda_env=None, registered_model_name=None, **kwargs):
+    Model.log(artifact_path=artifact_path, flavor=mlflow.torchscript, model=model,
+              conda_env=conda_env, registered_model_name=registered_model_name, **kwargs)
+
+
+def save_model(model, path, conda_env=None, mlflow_model=Model(), **kwargs):
+    if not hasattr(model, 'state_dict') or not hasattr(model, 'save'):
+        # If it walks like a duck and it quacks like a duck, then it must be a duck
+        raise TypeError("Argument 'model' should be a TorchScript model")
```

Comment:
Is there a type we can explicitly check against, vs checking attributes of the class? Just in case e.g. similar attributes (e.g. `state_dict`) are added to other classes in Pytorch?

Reply:
There are multiple types associated with the `jit`ed models (it's possible with `script` ing, `trace`ing and to make the things complicated, users can also make `script`ed python functions)

Reply:
I think we have a relatively (than I thought) easier way to check this. Updated the code

Reply:
Hm sorry, I think because of the force push I can't see the commits where you added the new check - looks like we're still checking attributes of the object rather than its type?